indexer/string_file_values.hpp

```diff
@@ -15,141 +20,262 @@
   {
-    rw::ReadVectorOfPOD(reader, m_value);
+    ReaderSource<TReader> src(reader);
+    DeserializeFromSource(src, v);
   }
 
-  inline void const * data() const { return m_value.data(); }
+  template <typename TSource>
+  void DeserializeFromSource(TSource & src, TValue & v) const
+  {
+    v.m_pt = serial::LoadPoint(src, m_codingParams);
+    v.m_featureId = ReadPrimitiveFromSource<uint32_t>(src);
+    v.m_rank = ReadPrimitiveFromSource<uint8_t>(src);
+  }
 
-  inline size_t size() const { return m_value.size() * sizeof(ValueT::value_type); }
+private:
+  serial::CodingParams m_codingParams;
+};
 
-  bool operator<(SerializedFeatureInfoValue const & value) const { return m_value < value.m_value; }
+template <>
+class SingleValueSerializer<FeatureIndexValue>
+{
+public:
+  using TValue = FeatureIndexValue;
+
+  SingleValueSerializer() = default;
 
-  bool operator==(SerializedFeatureInfoValue const & value) const
+  // todo(@mpimenov). Remove.
+  SingleValueSerializer(serial::CodingParams const & /* codingParams */) {}
+
+  // The serialization and deserialization is needed for StringsFile.
+  // Use ValueList for group serialization in CBVs.
+  template <typename TWriter>
+  void Serialize(TWriter & writer, TValue const & v) const
   {
-    return m_value == value.m_value;
+    WriteToSink(writer, v.m_featureId);
   }
 
-  void swap(SerializedFeatureInfoValue & value) { m_value.swap(value.m_value); }
+  template <typename TReader>
+  void Deserialize(TReader & reader, TValue & v) const
+  {
+    ReaderSource<TReader> src(reader);
+    DeserializeFromSource(src, v);
+  }
 
-  ValueT m_value;
+  template <typename TSource>
+  void DeserializeFromSource(TSource & src, TValue & v) const
+  {
+    v.m_featureId = ReadPrimitiveFromSource<uint64_t>(src);
+  }
 };
 
-/// This template is used to accumulate and serialize a group of
-/// values of the same type.
-template <typename Value>
+// This template is used to accumulate, serialize and deserialize
+// a group of values of the same type.
+template <typename TValue>
 class ValueList;
 
-/// ValueList<FeatureIndexValue> serializes a group of features
-/// indices as a compressed bit vector, thus, allowing us to save a
-/// disk space.
+// ValueList<FeatureIndexValue> serializes a group of feature
+// indices as a compressed bit vector.
 template <>
 class ValueList<FeatureIndexValue>
 {
 public:
-  void Append(FeatureIndexValue const & value)
-  {
-    // External-memory trie adds <string, value> pairs in a sorted
-    // order, thus, values are supposed to be accumulated in a
-    // sorted order, and we can avoid sorting them before construction
-    // of a CompressedBitVector.
-    ASSERT(m_offsets.empty() || m_offsets.back() <= value.m_value, ());
-    if (!m_offsets.empty() && m_offsets.back() == value.m_value)
-      return;
-    m_offsets.push_back(value.m_value);
+  using TValue = FeatureIndexValue;
+
+  ValueList() : m_cbv(unique_ptr<coding::CompressedBitVector>()) {}
```

Comment:
No need to initialize m_cbv - it's default ctor does exactly what you're going to achieve via this code.